LinkControl - adds search results label for initial suggestions

getdave opened the pull request:
As per https://github.com/WordPress/gutenberg/issues/19513, this PR adds a label to the search results for initial suggestions.

**Note**: this does not attempt to address [this feedback](https://github.com/WordPress/gutenberg/issues/19513#issuecomment-573070658).

> Could the URL protocol be dropped?

This will be handled in another PR.

## How has this been tested?
* Unit tests
* Manual testing:

To test:

* Add Nav Block
* See initial search suggestions
* See "Recently modified" above search results.



## Screenshots <!-- if applicable -->

<img width="623" alt="Screenshot 2020-01-15 at 11 54 39" src="https://user-images.githubusercontent.com/444434/72431938-d57ee100-378d-11ea-9198-87b46180e7ed.png">

## Types of changes
New feature (non-breaking change which adds functionality)
## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

getdave committed:
Adds search results label for initial suggestions

getdave committed:
Conditional render visually hidden label for normal search results

frontdevde reviewed (approved):
Tested as per instructions and it worked as expected. Code looks good as well 👍
Props for using the addition of the label to improve the a11y, nice one! 👏

jeryj commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -131,16 +131,25 @@ export function LinkControl( {
-				<div { ...suggestionsListProps } className={ resultsListClasses }>
+				{ ! isInitialSuggestions ? <VisuallyHidden>{ SearchResultsLabel }</VisuallyHidden> : SearchResultsLabel }
+
+				<div { ...suggestionsListProps } className={ resultsListClasses } aria-labelledby={ searchResultsLabelId }>
```
Nice job on adding `aria-labelledby`! 🙌 

I think there may be a slight difference on using that on visible vs hidden labels: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role

I'm not positive I'm fully understanding the implementation though: 
> **aria-label**
A human-readable string value which identifies the listbox.  If there's a visible label, then aria-labelledby should be used instead to refer to that label.

> **aria-labelledby**
Identifies the visible element or elements in a space-separated list of element IDs which identify the listbox.  If there's no visible label, then aria-label should be used instead to include a label.  (Note: "labelled", with two L's, is the correct spelling based on the accessibility API conventions.)

getdave replied:
Good catch but I don't follow the a11y guidelines here. Do I need to just add `aria-label` on the element when it's not visible?

shaunandrews commented:
Looks good but I question the use of the word “modified”. I think “Recently updated” was more approachable — modified feels strange in this context.

getdave committed:
Tweak wording from “modified” to “updated”

Addresses https://github.com/WordPress/gutenberg/pull/19665#issuecomment-574687453

jeryj commented on packages/block-editor/src/components/link-control/style.scss:
```diff
@@ -67,6 +67,12 @@
+.block-editor-link-control__search-results-label {
+	padding: 15px 30px 0 30px;
+	display: block;
+	font-size: 1.1em;
```
It seems like all the other `font-size` definitions are in `px`. Should this be changed to align with that?

getdave replied:
I want it to be based on a scale so it's "slightly larger" than whatever the base font size is for the component.

jeryj commented on packages/block-editor/src/components/link-control/test/index.js:
```diff
@@ -329,7 +329,10 @@ describe( 'Default search suggestions', () => {
+		const searchResultsWrapper = container.querySelector( '[role="listbox"]' );
+		const initialSearchResultElements = searchResultsWrapper.querySelectorAll( '[role="option"]' );
+
+		const searchResultsLabel = container.querySelector( `#${ searchResultsWrapper.getAttribute( 'aria-labelledby' ) }` );
```
If something did get changed with `aria-labelledby`, this may need to be changed as well.

getdave replied:
Tests pass because this attr is available when dealing with initial suggestions. So we should be all good!

getdave committed:
Fixes unecessary boolean inversion.

getdave commented:
@shaunandrews I've updated to "updated" 😆. On that basis are you happy to 👍 ?

getdave committed:
Refactor arial-labelled by and aria-label implementation to confirm to spec

https://www.w3.org/TR/wai-aria/#aria-label

getdave committed:
Use undefined instead of boolean false to avoid rendering of id and aria-labelledby

shaunandrews reviewed (approved):
🚢
